### PR TITLE
[LLVM][NVPTX] Add NVPTX codegen support for fence.proxy.tensormap

### PR DESCRIPTION
**Pull request by schwarzschild-radius**
This commit adds LLVM Intrinsics and NVPTX codegen support for `fence.proxy.tensormap` with lit tests under fence-proxy-tensormap.ll. Also, added Intrinsics documentation in NVPTXUsage.rst

### CODE REVIEW

**Comment by durga4github** [comment]
Changes look good to me. Requested review from Artem,

**Comment by durga4github** [comment]
@Artem-B , Could you please review this change?

File: llvm/include/llvm/IR/IntrinsicsNVVM.td
Comment (schwarzschild-radius) [reply llvm/include/llvm/IR/IntrinsicsNVVM.td]: > I'm not familiar enough with the new synchronization instructions, so I can't tell you what is the right way to specify their behavior in LLVM, but it's not uncommon for various barrier instructions to come with a IntrHasSideEffects to make sure that LLVM would not move them around.

By default intrinsics are assumed to be side-effecting unless specified otherwise (https://github.com/llvm/llvm-project/blob/67730ae19c6bcb08dca292e0576b6cd55a843932/llvm/include/llvm/IR/Intrinsics.td#L168). Added `IntrArgMemOnly` attribute to relax constraints for the acquire variants

> Also, are there any concerns regarding replicating/merging these instructions across different control flow paths? Do they need IntrConvergent ?

fence.* variants don't require `IntrConvergent` as they can be executed divergently as well

File: llvm/test/CodeGen/NVPTX/fence-proxy-tensormap.ll
Comment (schwarzschild-radius) [reply llvm/test/CodeGen/NVPTX/fence-proxy-tensormap.ll]: Fixed the ptxas version in the latest revision

File: llvm/test/CodeGen/NVPTX/fence-proxy-tensormap.ll
Comment (schwarzschild-radius) [reply llvm/test/CodeGen/NVPTX/fence-proxy-tensormap.ll:24]: Updated the ISel pattern to just match constant integer 128 argument and added a verifier check to emit diagnostics for other values

File: llvm/docs/NVPTXUsage.rst
Comment (gonzalobg) [inline llvm/docs/NVPTXUsage.rst:279]: I think this needs to mention the direction of the cross-proxy uni-directional fences: from which proxy to which proxy. In this case, these cross-proxy-fence from generic-proxy to tensormap-proxy.

For documenting which intrinsics map to which PTX instructions, a table would be clearer.

Comment (gonzalobg) [inline llvm/docs/NVPTXUsage.rst:279]: nit: missing a colon 

```suggestion
The '``@llvm.nvvm.fence.proxy.tensormap.release.*``' intrinsic emits ``fence.proxy.tensormap::generic.release.*`` and '``@llvm.nvvm.fence.proxy.tensormap.acquire.*``' intrinsic emits ``fence.proxy.tensormap::generic.acquire.* [addr], size;``. ``nvvm.fence.proxy.tensormap*`` is a uni-directional fence used to establish ordering between memory accesses that may happen through different proxies. ``nvvm.fence.proxy.tensormap.release`` can form a release sequence that synchronizes with an acquire sequence that contains the ``nvvm.fence.proxy.tensormap.acquire`` proxy fence.
```

File: llvm/docs/NVPTXUsage.rst
Comment (gonzalobg) [inline llvm/docs/NVPTXUsage.rst:281]: nit: half-open range?

```suggestion
The address operand ``addr`` and the operand ``size`` together specifies the memory range ``[addr, addr+size)`` on which the ordering guarantees on the memory accesses across the proxies is to be provided. The only supported value for the ``size`` operand is ``128`` and must be an immediate. For more information, see `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-membar>`_.
```

Comment (gonzalobg) [inline llvm/docs/NVPTXUsage.rst:281]: I think this should call out that address must either be `global` or `generic` pointing to global.

Do we have a way to annotate an intrinsic with that, to help addrspace propagation? (i.e. such that if someone uses this intrinsic on a pointer, we propagate through out that the pointer, and any pointer it was derived from / derived from it, must point to global.

File: llvm/docs/NVPTXUsage.rst
Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Added `generic` explicitly in the intrinsic to mark the direction of the fencing

File: llvm/docs/NVPTXUsage.rst
Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Added documentation to explicitly mention the proxy direction from generic to tensormap

Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Added a `colon` in the latest revision

File: llvm/docs/NVPTXUsage.rst
Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Fixed the `half open` range and typo

Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Updated the documentation to specify that addr must either be global or generic point to global

> Do we have a way to annotate an intrinsic with that, to help addrspace propagation? That is, such that if someone uses this intrinsic on a pointer, we propagate throughout that the pointer, and any pointer it was derived from / derived from it, must point to global?

I don't think we have an annotation for addrspace propagation. The `inferAddrspace` pass takes cares of addrspace inference but in this case, the instruction itself takes a generic address and does not have a specialized version for global addrspace,  so we can use generic addrspace pointers to avoid redundancy

File: llvm/docs/NVPTXUsage.rst
Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: Replace `tensormap.generic` with `tensormap_generic` to explicitly state `to_proxy::from_proxy`  and also differentiate generic-addrspace and generic-proxy

**Comment by gonzalobg** [comment]
Thanks Pradeep, this LGTM.

**Review by gonzalobg — APPROVED** [review]

**Comment by schwarzschild-radius** [comment]
@Artem-B I have addressed all of your comments, please take a look when you get a chance

**Review by Artem-B — APPROVED** [review]
LGTM with a doc nit.

File: llvm/docs/NVPTXUsage.rst
Comment (gonzalobg) [inline llvm/docs/NVPTXUsage.rst:279]: The way proxies are defined in PTX subtly differs from how they were explored in the ISCA paper, and the PTX docs about them are still in early days and evolving, so my preference is to just link the first use of proxy as follows, to avoid the LLVM NVPTX backend docs getting out of sync:

```suggestion
The ``@llvm.nvvm.fence.proxy.tensormap_generic.*`` is a uni-directional fence used to establish ordering between a prior memory access performed via the generic `proxy<https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#proxies>_` and a subsequent memory access performed via the tensormap proxy. ``nvvm.fence.proxy.tensormap_generic.release`` can form a release sequence that synchronizes with an acquire sequence that contains the ``nvvm.fence.proxy.tensormap_generic.acquire`` proxy fence. The following table describes the mapping between LLVM Intrinsic and the PTX instruction:
```

Comment (Artem-B) [reply llvm/docs/NVPTXUsage.rst]: The change landed with no definition of what a proxy is, and @gonzalobg's suggestion was resolved w/o applying it.

I'm not sure if that was intentional or an oversight.

**Comment by Artem-B** [comment]
@durga4github -- please hold off on merging the changes while the review conversation is still in progress.

File: llvm/docs/NVPTXUsage.rst
Comment (schwarzschild-radius) [reply llvm/docs/NVPTXUsage.rst]: I did not add a definiton to proxy because Gonzalo explained that the PTX definition is evolving and he suggested us to just add a link to the PTX's proxy defintion and I have included Gonzalo's suggestion in the commit (https://github.com/llvm/llvm-project/blob/cf416e0a1bd1b3ff6edc662116295f6fffe24be0/llvm/docs/NVPTXUsage.rst?plain=1#L279). Please let me know if I am missing something

Comment (gonzalobg) [reply llvm/docs/NVPTXUsage.rst]: I think this should have waited until Artem finished their review before merging it, since they had started reviewing this change but hadn't approved it yet.
Now its too late too late for that, but things like this happen, and can be fixed: @Artem-B we can continue the review here, and once its finished, do the changes as a follow up. How does that sound?

Comment (Artem-B) [reply llvm/docs/NVPTXUsage.rst]: @schwarzschild-radius My apologies, I do see the link to the proxy docs you've added. I must've been looking at the older revision of the patch. Github's review UI is still surprising me at the least convenient moments. :-/

**Comment by durga4github** [comment]
Sure, I saw Pradeep had added the links and you had approved; so clicked the merge button.
I will wait for the author/reviewer to resolve all conversations before merging,

**Pull request merged**
